Make package files stand alone for type checking

trws opened the pull request:
After some chats with @tgamblin, here's a first cut at getting all the package files to type check.  There's more work to do for the dynamically poked names like `make` and `spack_cxx`, but switching the name, adding a star import to every package file, and removing the vestigial `from spack import *`.  The rope script I put together for this also ran optimize imports, so the imports are sorted and split into groups by whether they're builtin or from the project.

After this change, we go from tens of thousands of pyright errors on the package repository to this:

```
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/3proxy/package.py
  27:9 - error: "make" is not defined (reportUndefinedVariable)
  30:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/abduco/package.py
  24:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/abi-compliance-checker/package.py
  25:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/abi-dumper/package.py
  29:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/abinit/package.py
  182:9 - error: "make" is not defined (reportUndefinedVariable)
  183:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/ace/package.py
  33:9 - error: "env" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/acpica-tools/package.py
  22:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/activeharmony/package.py
  28:9 - error: "make" is not defined (reportUndefinedVariable)
  32:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/adios/package.py
  152:13 - error: "env" is not defined (reportUndefinedVariable)
  153:13 - error: "env" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/admixtools/package.py
  43:13 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/alglib/package.py
.  ges/apcomp/package.py
  95:24 - error: "env" is not defined (reportUndefinedVariable)
  60:35 - error: "std_cmake_args" is not defined (reportUndefinedVariable)
  62:28 - error: "std_cmake_args" is not defined (reportUndefinedVariable)
  67:13 - error: "cmake" is not defined (reportUndefinedVariable)
  69:13 - error: "make" is not defined (reportUndefinedVariable)
  71:13 - error: "make" is not defined (reportUndefinedVariable)
  84:22 - error: "env" is not defined (reportUndefinedVariable)
  85:24 - error: "env" is not defined (reportUndefinedVariable)
  94:26 - error: "env" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/ape/package.py
  44:9 - error: "configure" is not defined (reportUndefinedVariable)
  45:9 - error: "make" is not defined (reportUndefinedVariable)
  46:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/apr-util/package.py
  99:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/aragorn/package.py
  22:25 - error: "spack_cc" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/argon2/package.py
  23:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/argp-standalone/package.py
  26:9 - error: "make" is not defined (reportUndefinedVariable)
  27:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/arpack-ng/package.py
  99:24 - error: "std_cmake_args" is not defined (reportUndefinedVariable)
  127:9 - error: "cmake" is not defined (reportUndefinedVariable)
  128:9 - error: "make" is not defined (reportUndefinedVariable)
  130:13 - error: "make" is not defined (reportUndefinedVariable)
  131:9 - error: "make" is not defined (reportUndefinedVariable)
  155:9 - error: "configure" is not defined (reportUndefinedVariable)
  156:9 - error: "make" is not defined (reportUndefinedVariable)
  158:13 - error: "make" is not defined (reportUndefinedVariable)
  159:9 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/ascent/package.py
  171:36 - error: "site_packages_dir" is not defined (reportUndefinedVariable)
  181:35 - error: "std_cmake_args" is not defined (reportUndefinedVariable)
  183:28 - error: "std_cmake_args" is not defined (reportUndefinedVariable)
  188:13 - error: "cmake" is not defined (reportUndefinedVariable)
  190:13 - error: "make" is not defined (reportUndefinedVariable)
  194:17 - error: "make" is not defined (reportUndefinedVariable)
  196:13 - error: "make" is not defined (reportUndefinedVariable)
  222:13 - error: "cmake" is not defined (reportUndefinedVariable)
  223:13 - error: "make" is not defined (reportUndefinedVariable)
  236:13 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/astra/package.py
  56:9 - error: "mkdir" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/astyle/package.py
  35:47 - error: "spack_cxx" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/athena/package.py
  160:27 - error: "spack_cc" is not defined (reportUndefinedVariable)
  161:28 - error: "spack_cc" is not defined (reportUndefinedVariable)
  268:9 - error: "mkdir" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/atk/package.py
  55:13 - error: "meson" is not defined (reportUndefinedVariable)
  55:26 - error: "std_meson_args" is not defined (reportUndefinedVariable)
  56:13 - error: "ninja" is not defined (reportUndefinedVariable)
  57:13 - error: "ninja" is not defined (reportUndefinedVariable)
  61:9 - error: "configure" is not defined (reportUndefinedVariable)
  62:9 - error: "make" is not defined (reportUndefinedVariable)
  64:13 - error: "make" is not defined (reportUndefinedVariable)
  65:9 - error: "make" is not defined (reportUndefinedVariable)
  67:13 - error: "make" is not defined (reportUndefinedVariable)
/Users/scogland1/Workspace/Projects/spack/spack/var/spack/repos/builtin/packages/atlas/package.py
  99:25 - error: "spack_cc" is not defined (reportUndefinedVariable)
  100:25 - error: "spack_f77" is not defined (reportUndefinedVariable)
  106:47 - error: "spack_cc" is not defined (reportUndefinedVariable)
  116:13 - error: "make" is not defined (reportUndefinedVariable)
  117:13 - error: "make" is not defined (reportUndefinedVariable)
  118:13 - error: "make" is not defined (reportUndefinedVariable)
  119:13 - error: "make" is not defined (reportUndefinedVariable)
  122:21 - error: "make" is not defined (reportUndefinedVariable)
```

trws committed:
rename spack.pkgkit to spack.std

trws committed:
update and optimize imports in builtin repo

trws committed:
update and optimize imports in builtin.mock repo

adamjstewart commented:
Another thing you'll want to update is the new package creation template in `lib/spack/spack/cmd/create.py`

adamjstewart commented:
Now that we're explicitly adding `from spack.std import *` to every package, do we plan on enforcing this? Should we remove the hack that automatically adds this to package files? Will that make Spack noticeably faster?

adamjstewart commented:
I don't love the name `spack.std`. @tgamblin would it be possible to merge that logic into `spack.package`? I realize that `spack.package` is already humongous, but most of that stuff gets imported into the `package.py` files anyway right?

trws commented:
Yes, it should probably be enforced.  I should change around the rules I added to the flake8 pr #20376 to do that, since I was already tweaking the rules for these imports.

We talked a bit about using `spack.package` or `spack.package.std`, but the former isn't really quite what you want, it pulls in things that might not be desirable and doesn't pull in things you'd want, the latter would work but is longer.  I don't really care what the name is so much as having the import there explicitly so tooling can work.

It probably won't make it noticeable, but it might.  If we can actually get rid of the need to modify the package, and the module after loading, then that might be noticeable because we could let it cache a compiled version.  The bigger performance win would be to automatically produce a loadable zip or pex from the repo and use that, but not planning to go there just now since the logic for when to update it would be kinda tricky.

trws commented:
Hm, I'll update the flake8 config here for that specifically so it can pass style, wasn't thinking about that.

trws commented:
Dangit, I'll also fix the flake8 command so it can actually work on this PR.  It makes an argument list that's too long for the system buffers.

trws committed:
make flake8 work with changes and long file lists

trws committed:
shift module level injected functions to std as possible, stub others

trws commented:
This last bit makes it possible to type check the common injected commands, it relies on the name injection overriding the names in `spack.std` when necessary.

This may be where this bottoms out without some much deeper work though.  There are two remaining limitations:

1. `setup_dependent_package`: Since the set of module injections can be entirely dynamic, and it isn't something like a class of known type masquerading as a module, I think the only way to deal with this may be to write a plugin for Mypy to resolve the names. I did this for flux so it could type-check the interfaces we auto-generate over there for wrapped c libraries, but it's a non-trivial amount of work, so if anyone can come up with even a stub solution we may be better off.
2. `spack.pkg.*`: I think Mypy can deal with this just by setting that these imports should be ignored, but upright can't do that based on the path of the thing to be imported, so we may have to turn off import checking for package checks unless we can figure something out.

trws committed:
remove dynamic addition to __all__ to fix doc gen

tgamblin commented:
> I don't love the name spack.std. @tgamblin would it be possible to merge that logic into spack.package? I realize that spack.package is already humongous, but most of that stuff gets imported into the package.py files anyway right?

We could also move `package` to `package_base` or something like that so that it's separate.  `spack.pkg` is already taken as it is the super-namespace for all repos and package modules -- that was also my first choice.

I do like `spack.package`, and it is an internal API so a bit easier to move.  Thoughts?

adamjstewart commented:
What is in `spack.package` that shouldn't be imported by every package? I feel like most of that stuff is needed in every package anyway.

trws commented:
I'd be fine with moving package to package_base or something then using package for this, or making this `spack.package.lib` or some such thing.  Directly pulling in package would pull in a lot of names that most packages don't need to have in scope, and might conflict with things we want to have in there.  If it weren't a star import it would be fine, but since it is, I don't think we want to directly use the existing package.py.

adamjstewart commented:
It seems to me like most of the stuff in `spack.package` is base classes, so the namespace wouldn't really get polluted. Also, we could define `__all__` to specify what gets imported by `import *`, right?

trws commented:
It's kinda not.  This is the list, as grabbed from the dict so there are some extra items in there.

```
['__name__',
 '__doc__',
 '__package__',
 '__loader__',
 '__spec__',
 '__file__',
 '__cached__',
 '__builtins__',
 'base64',
 'collections',
 'contextlib',
 'copy',
 'functools',
 'hashlib',
 'inspect',
 'os',
 're',
 'shutil',
 'sys',
 'textwrap',
 'time',
 'traceback',
 'six',
 'types',
 'fsys',
 'tty',
 'spack',
 'fs',
 'memoized',
 'LinkTree',
 'OrderedDict',
 'YamlFilesystemView',
 'PackageInstaller',
 'InstallError',
 'TestFailure',
 'TestSuite',
 'which',
 'ProcessError',
 'Prefix',
 'stage_prefix',
 'Stage',
 'ResourceStage',
 'StageComposite',
 'package_hash',
 'Version',
 '_ALLOWED_URL_SCHEMES',
 '_spack_build_logfile',
 '_spack_build_envfile',
 '_spack_configure_argsfile',
 'InstallPhase',
 'detectable_packages',
 'DetectablePackageMeta',
 'PackageMeta',
 'run_before',
 'run_after',
 'on_package_attributes',
 'PackageViewMixin',
 'test_log_pathname',
 'PackageBase',
 'test_process',
 'inject_flags',
 'env_flags',
 'build_system_flags',
 'BundlePackage',
 'Package',
 'install_dependency_symlinks',
 'use_cray_compiler_names',
 'flatten_dependencies',
 'possible_dependencies',
 'FetchError',
 'PackageStillNeededError',
 'PackageError',
 'PackageVersionError',
 'NoURLError',
 'InvalidPackageOpError',
 'ExtensionError',
 'ActivationError',
 'DependencyConflictError']
```

We could certainly define __all__, but I'm not sure the list of things used by other parts of spack would mesh with what we would want package to expose to packages.

adamjstewart commented:
Meh, those are almost entirely classes and imports. I'm fine with whatever we decide, but `spack.std` is a terrible name.

tgamblin commented:
I think it would be good to make `from spack.package import *` happen.

The whole reason we're here, though, is because we originally did `from spack import *` in all packages, and then we had to add some magic in the importer to undo that because imports were *really slow*.  I think getting a sufficiently encapsulated namespace here is key to ensuring that we can solve slow import problems as Spack evolves.  Even if you think that what's in `package` *now* is ok to import, that doesn't mean we won't add something later that causes all of Spack to slow to a crawl.  I don't think we should constrain what can be imported by core developers by what's needed by packages or vice versa.

So I'm thinking renaming `package.py` would be best, as I agree that `spack.std` is pretty horrible, and it would be nice if the import for packages made sense.  The only issue I see is that this would break external consumers of the package API, e.g. those who use [Spack extensions](https://spack.readthedocs.io/en/latest/extensions.html) or write [scripts](https://spack-tutorial.readthedocs.io/en/latest/tutorial_spack_scripting.html).

We never promised the API would be stable (and we're at `0.x`) so there is that.  This is a small-ish change (`import spack.package` ->  `import spack.package_base`) but might still be good to warn some folks.

Thinking @gartung, @chissg, @mrmundt, maybe @mwkrentel but I don't think he uses any of that.  Any others? Do any of you all import `spack.package` in your code?  @alalazo do you know of anyone else we should consider pinging?

mrmundt commented:
> I think it would be good to make `from spack.package import *` happen.
> 
> The whole reason we're here, though, is because we originally did `from spack import *` in all packages, and then we had to add some magic in the importer to undo that because imports were _really slow_. I think getting a sufficiently encapsulated namespace here is key to ensuring that we can solve slow import problems as Spack evolves. Even if you think that what's in `package` _now_ is ok to import, that doesn't mean we won't add something later that causes all of Spack to slow to a crawl. I don't think we should constrain what can be imported by core developers by what's needed by packages or vice versa.
> 
> So I'm thinking renaming `package.py` would be best, as I agree that `spack.std` is pretty horrible, and it would be nice if the import for packages made sense. The only issue I see is that this would break external consumers of the package API, e.g. those who use [Spack extensions](https://spack.readthedocs.io/en/latest/extensions.html) or write [scripts](https://spack-tutorial.readthedocs.io/en/latest/tutorial_spack_scripting.html).
> 
> We never promised the API would be stable (and we're at `0.x`) so there is that. This is a small-ish change (`import spack.package` -> `import spack.package_base`) but might still be good to warn some folks.
> 
> Thinking @gartung, @chissg, @mrmundt, maybe @mwkrentel but I don't think he uses any of that. Any others? Do any of you all import `spack.package` in your code? @alalazo do you know of anyone else we should consider pinging?

@tgamblin Personally speaking, I never import `spack.package`. The collection of my `spack` imports are as follows:

```
import llnl.util.tty as tty
from spack.main import SpackCommand # for various uses
import spack.environment as spenv
```
So I don't think this change will affect anything on my end.

trws committed:
rename spack.package to spack.package_base

First part of refactor to allow std to become spack.package

trws committed:
fix more references to spack.package

trws committed:
add vermine ignore for typing in std

trws committed:
rename std to package

This also fixes some formatting errors that didn't come to light until
the style update landed, and adds an import unused exception to flake8
for spack.package.

trws committed:
missed two uses of spack.std

trws committed:
remove typing import from kripke module

trws committed:
type annotations for build systems

trws committed:
make incorrectly instance method static

trws committed:
move fedora_patch out of class in Tcsh package

trws committed:
fix style issues

adamjstewart commented:
Another possible fix for the name collision between `package` and `package_base`: we could move one or the other into `lib/spack/spack/build_systems/`

adamjstewart commented on lib/spack/spack/repo.py:
```diff
@@ -69,14 +69,14 @@ def get_full_namespace(namespace):
 #:
 #: TODO: At some point in the future, consider removing ``from spack import *``
-#: TODO: from packages and shifting to from ``spack.pkgkit import *``
-_package_prepend = 'from spack.pkgkit import *'
+#: TODO: from packages and shifting to from ``spack.package import *``
```
This TODO can be removed now, since that's exactly what this PR does.

trws replied:
Ah, quite so, I'll pull that out.

adamjstewart commented on lib/spack/spack/repo.py:
```diff
@@ -69,14 +69,14 @@ def get_full_namespace(namespace):
 #: core the top-level namespace polluted by package symbols this way.  To
 #: solve this, the top-level ``spack`` package contains very few symbols
 #: of its own, and importing ``*`` is essentially a no-op.  The common
-#: routines and directives that packages need are now in ``spack.pkgkit``,
+#: routines and directives that packages need are now in ``spack.package``,
 #: and the import system forces packages to automatically include
 #: this. This way, old packages that call ``from spack import *`` will
 #: continue to work without modification, but it's no longer required.
 #:
 #: TODO: At some point in the future, consider removing ``from spack import *``
-#: TODO: from packages and shifting to from ``spack.pkgkit import *``
-_package_prepend = 'from spack.pkgkit import *'
+#: TODO: from packages and shifting to from ``spack.package import *``
+_package_prepend = 'from spack.package import *'
```
I would be in favor of removing this logic entirely and forcing packages to include `from spack.package import *`. At the very least, we could deprecate it if we're worried about non-builtin packages. How much of a performance improvement will this give us?

trws replied:
Me too, but the idea was to leave it deprecated for a couple of versions for packages outside of the builtin repo.

I'm not sure on performance.  It depends on how much other stuff we can refactor out with the prepend removed I think, since the prepend is not super slow in and of itself, but having to load the whole thing and modify it rather than direct load it probably has a non-trivial cost, especially in that it can't be cached.

adamjstewart commented:
When I run `spack style` on develop, I see:
```
==> style: running mypy checks on spack.
var/spack/repos/builtin/packages/python/package.py: error: Duplicate module named 'package' (also at 'var/spack/repos/builtin/packages/apple-libuuid/package.py')
Found 1 error in 1 file (checked 2 source files)
```
Does this PR fix that, or is there another smaller, easier-to-merge PR that fixes that?

trws commented:
> When I run `spack style` on develop, I see:
> 
> ```
> ==> style: running mypy checks on spack.
> var/spack/repos/builtin/packages/python/package.py: error: Duplicate module named 'package' (also at 'var/spack/repos/builtin/packages/apple-libuuid/package.py')
> Found 1 error in 1 file (checked 2 source files)
> ```
> 
> Does this PR fix that, or is there another smaller, easier-to-merge PR that fixes that?

That should not be happening, the package's package.py files should be ignored by the config on develop.  it should be ignored by the `[mypy]` block at the start.  Could you send me the mypy version you have and double-check that the `.mypy.ini` matches this?  If it's causing problems I can add a small PR to exclude packages from mypy the way this PR does, but with the version I'm using it wasn't necessary so I left it out.

```ini
[mypy]
python_version = 3.7
files=lib/spack/llnl/**/*.py,lib/spack/spack/**/*.py
mypy_path=bin,lib/spack,lib/spack/external,var/spack/repos/builtin
# This and a generated import file allows supporting packages
namespace_packages=True
# To avoid re-factoring all the externals, ignore errors and missing imports
# globally, then turn back on in spack and spack submodules
ignore_errors=True
ignore_missing_imports=True

[mypy-spack.*]
ignore_errors=False
ignore_missing_imports=False
```

trws commented:
> Another possible fix for the name collision between `package` and `package_base`: we could move one or the other into `lib/spack/spack/build_systems/`

We could, I wouldn't mind doing that necessarily actually, but we would still get conflicts with all the package.py files in the packages sadly.

adamjstewart commented:
```console
$ mypy --version
mypy 0.740
```
`.mypy.ini`:
```ini
[mypy]
python_version = 3.7
files=lib/spack/llnl/**/*.py,lib/spack/spack/**/*.py
mypy_path=bin,lib/spack,lib/spack/external,var/spack/repos/builtin
# This and a generated import file allows supporting packages
namespace_packages=True
# To avoid re-factoring all the externals, ignore errors and missing imports
# globally, then turn back on in spack and spack submodules
ignore_errors=True
ignore_missing_imports=True

[mypy-spack.*]
ignore_errors=False
ignore_missing_imports=False

[mypy-packages.*]
ignore_errors=False
ignore_missing_imports=False

[mypy-llnl.*]
ignore_errors=False
ignore_missing_imports=False

[mypy-spack.test.packages]
ignore_errors=True

# ignore errors in fake import path for packages
[mypy-spack.pkg.*]
ignore_errors=True
ignore_missing_imports=True

# jinja has syntax in it that requires python3 and causes a parse error
# skip importing it
[mypy-jinja2]
follow_imports=skip
```

adamjstewart commented:
Now I'm having trouble reproducing. Ignore me for now, I'll open an issue after this gets merged if I encounter any problems.

bernhardkaindl commented:
@trws I assume this is obsolete now?

adamjstewart commented:
@haampie was talking about doing this on Slack, but it would probably be a new PR since this one is so out-of-date. Let's close this and we can create a new PR.

trws commented:
*sigh* yeah, I really want to come back around and see this done somehow.  @haampie, if you'd like to chat sometime or want the rope scripts poke me.  I'd love to see this happen and just ran out of time to spend on it sadly.

adamjstewart commented:
I think the best way forward is to do things incrementally:

1. Update `lib/spack/spack/cmd/create.py` to add the new import line
2. Update all existing packages with the new import line
3. Remove the import line being inserted during import

If those are separate PRs they should be quick and easy to review? The rest of the stuff is where we got bogged down last time if I remember correctly.
